arangod/Aql/SortBlock.cpp

```diff
@@ -21,256 +21,616 @@
-  std::vector<std::pair<uint32_t, uint32_t>> coords;
-  coords.reserve(sum);
+  ~ConstrainedHeapSorter() { releaseHeapBuffer(); }
 
-  // install the coords
-  // we are intentionally using uint32_t here to save memory and
-  // have better cache utilization
-  uint32_t count = 0;
+  std::pair<arangodb::aql::ExecutionState, arangodb::Result> fetch() override {
+    using arangodb::aql::AqlItemBlock;
+    using arangodb::aql::ExecutionBlock;
+    using arangodb::aql::ExecutionState;
 
-  for (auto const& block : _buffer) {
-    uint32_t const n = static_cast<uint32_t>(block->size());
+    ExecutionState res = ExecutionState::HASMORE;
+    // suck all blocks through _buffer into heap
+    while (res != ExecutionState::DONE) {
+      res = _fetch(ExecutionBlock::DefaultBatchSize()).first;
+      if (res == ExecutionState::WAITING) {
+        return {res, TRI_ERROR_NO_ERROR};
+      }
 
-    for (uint32_t i = 0; i < n; i++) {
-      coords.emplace_back(std::make_pair(count, i));
+      // handle batch
+      while (!_buffer.empty()) {
+        std::unique_ptr<AqlItemBlock> src{_buffer.front()};
+        auto cleanup =
+            arangodb::scopeGuard([this]() -> void { _buffer.pop_front(); });
+        ensureHeapBuffer(src.get());  // make sure we have a dst
+        for (size_t row = 0; row < src->size(); row++) {
+          pushRow(row);
+        }
+      }
     }
-    ++count;
+
+    return {ExecutionState::DONE, TRI_ERROR_NO_ERROR};
   }
 
-  // comparison function
-  OurLessThan ourLessThan(_trx, _buffer, _sortRegisters);
+  arangodb::Result sort() override {
+    using arangodb::aql::AqlItemBlock;
+    using arangodb::aql::AqlValue;
+    using arangodb::aql::ExecutionBlock;
+    using arangodb::aql::RegisterId;
+    using arangodb::basics::catchVoidToResult;
 
-  // sort coords
-  if (_stable) {
-    std::stable_sort(coords.begin(), coords.end(), ourLessThan);
-  } else {
-    std::sort(coords.begin(), coords.end(), ourLessThan);
-  }
+    TRI_IF_FAILURE("SortBlock::doSorting") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_DEBUG);
+    }
 
-  // here we collect the new blocks (later swapped into _buffer):
-  std::deque<AqlItemBlock*> newbuffer;
+    // make sure we don't have less than limit
+    size_t total = std::min(_limit, _rowsPushed);
+    uint32_t count = 0;
 
-  try {  // If we throw from here, the catch will delete the new
-    // blocks in newbuffer
+    // sort rows
+    std::sort(_rows.begin(), _rows.end(), _cmpHeap);
 
+    // here we collect the new blocks:
+    TRI_ASSERT(_buffer.empty());
     count = 0;
-    RegisterId const nrRegs = _buffer.front()->getNrRegs();
-
+    RegisterId const nrRegs = _heapBuffer.front()->getNrRegs();
     std::unordered_map<AqlValue, AqlValue> cache;
 
     // install the rearranged values from _buffer into newbuffer
+    while (count < total) {
+      size_t sizeNext = (std::min)(total - count, ExecutionBlock::DefaultBatchSize());
+      AqlItemBlock* next = _allocate(sizeNext, nrRegs);
```

Comment:
This and the following looks a bit overengineered...
Why not capture `next` in an `std::unique_ptr` instead, remove all the try...catch stuff and release `next` in case we want to transfer ownership?